### ISSUE TITLE
[Discussion] ISTIO and TFServing

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#464]
We'd like to integrate ISTIO with TFServing to enforce things such as

- Auth
- Quota policies.

**Comment by lluunn** [comment]
I will look into this

**Comment by jlewi** [comment]
@lluunn How's this going? Can you provide an update?

**Comment by jlewi** [comment]
I think it might be better to collect metrics from Envoy directly in front of TF serving rather than relying on Ambassador for that.

Before going down the route of Ambassador; could we get a PR submitted adding envoy as a sidecar to TFServing?

What else is required to make ISTIO work? How are you deploying the ISTIO components? And what about the Grafana dashboards?

I think it would be great if we could get the minimal ksonnet changes needed to support this committed. Then it would be great if we could use this approach to surface RPC metrics for our GH issue summarization example (note that example is actually using Seldon not TFServing); but we also have a web app in that case.

cc'ing some people who might have already tackled the problem of reporting RPC metrics to see if they have anything to  add.

/cc @cfegly @cliveseldon

**Comment by lluunn** [comment]
@ddysher I agree that we may not need ambassador after istio integration.
But I don't understand "Equipping only the serving service is not that much useful ". 
IMO, istio features such as routing, rollout, metrics, etc all make sense only to serving parts of kubeflow. Is there any use case that you need istio for other kubeflow components?

**Comment by lluunn** [comment]
@cliveseldon 
Yeah I think we want to roll out/roll back, canary, A/B testing for TF Serving, which istio seems to be a natural fit.

**Comment by ddysher** [comment]
@lluunn  Agreed that the features make sense for serving parts w.r.t kubeflow; but in a larger picture, serving might just be part of an application stack - there can be cache, business logic, data processing, etc.  How to equip those services to unleash the full potential of service mesh is what i was trying to say.

**Comment by jlewi** [comment]
@ddysher My expectation is that K8s & Istio will eventually have a good story for north-south ( traffic in/out of cluster) as well as east-west (within cluster).

So I think in the near term we can use ISTIO for east-west traffic and continue to use Ambassador and some other improvisation for north south. Our GH issue is a good example because we can use ISTIO in between our Web App and the Seldon model. 

Right now I don't think K8s & ISTIO provide some of the north-south traffic features we want e.g.

- JWT validation
- External Auth
- Dynamic ingress routes
- Ingress management

In some of my conversations with members of ISTIO team some of these features are coming to ISTIO but aren't present yet. For example JWT evaluation is in envoy but I don't know if that's been upstreamed yet. In many cases, I think functionality is there but its not integrated with the control plane.

I think K8s is improving the ingress story as well for example, there is a [ServiceExtension CRD](
https://github.com/kubernetes/ingress-gce/pull/163) coming. I think ServiceExtensions will live side by side with K8s services and potentially provide additional information.

/cc @kflynn @richarddli

**Comment by richarddli** [comment]
Currently Istio ingress only does path based routing for HTTP requests. Nothing more, nothing less. Istio (and Kube ingress in general) doesn't provide any sort of auth mechanism. Kubeflow also currently relies on URL rewrite, which is also something not supported in Kube ingress & Istio. There is discussion on the Istio API management group about these kinds of things, but as far as I can tell this is still a "futures" kind of conversation.

I'm not particularly familiar with the Kubeflow topology, so pardon me for my ignorance here, but I'm trying to understand what are the problem(s) Kubeflow has that Istio would solve?

If you have a shallow topology (e.g., ingress -> multiple Kubeflow services), Ambassador may be sufficient (it provides the observability, security, and canary routing capabilities here that's mentioned). If you have a deeper topology (e.g., ingress -> Kubeflow service 1 -> Kubeflow service 2 -> KF servce 3) that's where you need something like Istio. 

As an aside, there are several issues open in Ambassador about improving integration with Istio; we're looking for a concrete user who has a specific need for some Istio functionality to collaborate with.

**Comment by ddysher** [comment]
@jlewi Thanks for the write up. Yes, the whole area is evolving quickly and is becoming more mature.  There are also solutions like linkerd, conduit, and I believe there will be more solutions (we have our own solutions as well).  For kubeflow; however, are we expecting it to handle north-south traffic, do we need JWT validation IN kubeflow?  In a world where everything is pluggable and swappable, building a total solution doesn't seem to scale well IMO.  Is this integration within the scope of kubeflow?

**Comment by jlewi** [comment]
@ddysher I think for Kubeflow we need to have ingress solutions that we can recommend. That said if users don't want to use those solutions they can customize their deployment. To your earlier point about using ISTIO for the application stack; I believe we are doing that. For example, in our GH Issue Example our application consists of web-app & TFServing and we enable ISTIO on both of those.

There's a lot in this thread. I think there are at least 3 different issues

1. How do we deploy/rollout models?
1. Can/should we use ISTIO to manage north/south traffic?
1. How to use ISTIO to manage east-west traffic (e.g. enforce quota etc...)

**Comment by jlewi** [comment]
@lluunn Can you summarize the current ISTIO integration for serving? Are there any missing features we should try to add as part of our 0.2 release?

**Comment by lluunn** [comment]
TF Serving component integrates with Istio for 2 features now: 
 - RPC type metrics, number of requests, latencies etc
 - routing rules. X% to version1, Y% to version2. So can do simple A/B testing and rollout.

Detail [here](https://github.com/kubeflow/kubeflow/blob/master/components/k8s-model-server/istio-integration.md)

I think it's good for now?
